Cics STARTBROWSE improvements

mm-broadcom opened the pull request:
Fixes and implements the STARTBROWSE CICS command.

## How Has This Been Tested?

Created and ran a number of valid and invalid tests for the given command based upon IBM's grammar.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Improved Startbrowse CICS rules

Improved the Startbrowse CICS rules, added checks on said rules.

mm-broadcom committed:
Fixed Checkstyle errors, utility functions

mm-broadcom committed:
Link for processing

mm-broadcom committed:
Add unit tests

mm-broadcom committed:
Removed unused import

Removed unused import to fix the checkstyle error.

mm-broadcom committed:
Switch to using CICSTestUtils

Switch to using CICSTestUtils

mm-broadcom committed:
Add checks for duplicates

Add checks for duplicates of possible tokens

mm-broadcom committed:
Switch to using the CICSLexer versus strings

mm-broadcom committed:
Allowed for PROCESS and its TYPE to be flexible.

Allowed for PROCESS and its TYPE to be flexible. Fixed test to reflect the new grammar.

mm-broadcom commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -725,15 +725,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
 cics_startbrowse_process: PROCESS (PROCESSTYPE cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
+cics_startbrowse_timer: TIMER cics_data_value (BROWSETOKEN cics_data_area | cics_handle_response)+;
+
+cics_startbrowse_activityid_process: (ACTIVITYID cics_data_value | PROCESS cics_data_value PROCESSTYPE cics_data_value);
```
Fixed in latest commit.

mm-broadcom commented:
@slavek-kucera Would it be possible for you to review this?

mm-broadcom commented:
@KutluOzel-b Could you review this?

KutluOzel-b commented:
@mm-broadcom Please rebase the branch from development

mm-broadcom commented:
@KutluOzel-b Done, had to fix a duplicated function error first and run the tests locally before pushing the fix.

KutluOzel-b commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -716,15 +716,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
-cics_startbrowse_container: CONTAINER (ACTIVITYID cics_data_value | PROCESS cics_data_value PROCESSTYPE cics_data_value |
-                            CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
+                  cics_startbrowse_process | cics_startbrowse_timer);
+cics_startbrowse_activity: ACTIVITY (cics_startbrowse_activityid_process | BROWSETOKEN cics_data_area | cics_handle_response)+;
```
```suggestion
cics_startbrowse_activity: (ACTIVITY | cics_startbrowse_activityid_process | BROWSETOKEN cics_data_area | cics_handle_response)+;
```
To honor flexibility of CICS translator

KutluOzel-b commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -716,15 +716,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
-                            CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
+                  cics_startbrowse_process | cics_startbrowse_timer);
+cics_startbrowse_activity: ACTIVITY (cics_startbrowse_activityid_process | BROWSETOKEN cics_data_area | cics_handle_response)+;
+cics_startbrowse_container: CONTAINER (cics_startbrowse_activityid_process | CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
```
```suggestion
cics_startbrowse_container: (CONTAINER | cics_startbrowse_activityid_process | CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
```
Flexibility

KutluOzel-b replied:
```suggestion
cics_startbrowse_container: (CONTAINER | cics_startbrowse_activityid_process | CHANNEL cics_data_value | (BROWSETOKEN | RETCODE) cics_data_area | cics_handle_response)+;
```
As per Chace stated 
![image](https://github.com/user-attachments/assets/d2251935-28af-485c-b7f8-a65407971bb2)

KutluOzel-b commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -716,15 +716,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
+cics_startbrowse_container: CONTAINER (cics_startbrowse_activityid_process | CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
 cics_startbrowse_event: EVENT (BROWSETOKEN cics_data_area | ACTIVITYID cics_data_value | cics_handle_response)+;
 cics_startbrowse_process: PROCESS (PROCESSTYPE cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
+cics_startbrowse_timer: TIMER cics_data_value (BROWSETOKEN cics_data_area | cics_handle_response)+;
```
```suggestion
cics_startbrowse_timer: (TIMER cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
```

KutluOzel-b commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -716,15 +716,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
+cics_startbrowse_activity: ACTIVITY (cics_startbrowse_activityid_process | BROWSETOKEN cics_data_area | cics_handle_response)+;
+cics_startbrowse_container: CONTAINER (cics_startbrowse_activityid_process | CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
 cics_startbrowse_event: EVENT (BROWSETOKEN cics_data_area | ACTIVITYID cics_data_value | cics_handle_response)+;
 cics_startbrowse_process: PROCESS (PROCESSTYPE cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
```
```suggestion
cics_startbrowse_process: (PROCESS | PROCESSTYPE cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
```

slavek-kucera replied:
I think that would cause parsing issues...

KutluOzel-b commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -716,15 +716,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
+                  cics_startbrowse_process | cics_startbrowse_timer);
+cics_startbrowse_activity: ACTIVITY (cics_startbrowse_activityid_process | BROWSETOKEN cics_data_area | cics_handle_response)+;
+cics_startbrowse_container: CONTAINER (cics_startbrowse_activityid_process | CHANNEL cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
 cics_startbrowse_event: EVENT (BROWSETOKEN cics_data_area | ACTIVITYID cics_data_value | cics_handle_response)+;
```
```suggestion
cics_startbrowse_event: (EVENT | BROWSETOKEN cics_data_area | ACTIVITYID cics_data_value | cics_handle_response)+;
```

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -102,15 +102,76 @@ public CICSOptionsCheckBaseUtility(
     return true;
   }
 
+  protected void checkHasMandatoryOptions(TerminalNode rules, ParserRuleContext ctx, String options) {
```
We don't need this. All cics commands are handled as if they were terminal node list to honor flexibility, I think it is better stick to that.

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -102,15 +102,76 @@ public CICSOptionsCheckBaseUtility(
+   * @param optionalContext - The rule that is optional
+   * @param options - String of the element that is required.
+   */
+  protected void checkHasRequiredOption(ParserRuleContext requiredContext, ParserRuleContext optionalContext, String options) {
```
No need to have any of the `checkHasRequiredOption` methods

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -102,15 +102,76 @@ public CICSOptionsCheckBaseUtility(
-          VisitorUtility.constructLocality(ctx, context),
-          "Missing required option: ",
-          options);
+              ErrorSeverity.ERROR,
```
Please revert the formatting changes as they have not related with this PR.

mm-broadcom replied:
This is exactly what the development branch has it as right now. I tried copy/pasting the code from there, enabled whitespace changes to show and did not see any changes in Git.

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,120 @@
+            put(CICSLexer.CONTAINER, ErrorSeverity.ERROR);
+            put(CICSLexer.CHANNEL, ErrorSeverity.ERROR);
+            put(CICSLexer.EVENT, ErrorSeverity.ERROR);
+            put(CICSLexer.TIMER, ErrorSeverity.ERROR);
```
Add RETCODE

KutluOzel-b commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsExciStartBrowseStatement.java:
```diff
@@ -14,36 +14,83 @@
-          + "       PROGRAM-ID. EXCISTMTTEST.\n"
+
+  // Main Building Blocks
+  private static final String BASE_TEXT =
```
Remove the redundant template as you are already using `CICSTestUtils.noErrorTest` or `CICSTestUtils.errorTest`

mm-broadcom replied:
Done in latest commit.

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    private void checkActivity(CICSParser.Cics_startbrowse_activityContext ctx) {
+        checkHasMandatoryOptions(ctx.BROWSETOKEN(), ctx, "BROWSETOKEN");
+    }
+
+    private void checkContainer(CICSParser.Cics_startbrowse_containerContext ctx) {
+        checkHasMandatoryOptions(ctx.BROWSETOKEN(), ctx, "BROWSETOKEN");
+        if (!ctx.CHANNEL().isEmpty()) {
+            checkHasIllegalOptions(ctx.cics_startbrowse_activityid_process(), "ACTIVITYID or PROCESS");
+        }
+    }
+
+    private void checkEvent(CICSParser.Cics_startbrowse_eventContext ctx) {
+        checkHasMandatoryOptions(ctx.BROWSETOKEN(), ctx, "BROWSETOKEN");
+    }
+
+    private void checkProcess(CICSParser.Cics_startbrowse_processContext ctx) {
+        checkHasMandatoryOptions(ctx.BROWSETOKEN(), ctx, "BROWSETOKEN");
+        checkHasMandatoryOptions(ctx.PROCESSTYPE(), ctx, "PROCESSTYPE");
+    }
+
+    private void checkTimer(CICSParser.Cics_startbrowse_timerContext ctx) {
+        checkHasMandatoryOptions(ctx.BROWSETOKEN(), ctx, "BROWSETOKEN");
+    }
+
+    private void checkActivityIdProcess(CICSParser.Cics_startbrowse_activityid_processContext ctx) {
+        checkMutuallyExclusiveOptions("ACTIVITYID or PROCESS", ctx.ACTIVITYID(), ctx.PROCESS());
```
Add test cases for these checks

mm-broadcom replied:
Done.

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Add RETCODE support

mm-broadcom committed:
Add RETCODE to duplicate check options

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

mm-broadcom committed:
Added more invalid test cases

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: Kutlu <104970275+KutluOzel-b@users.noreply.github.com>

KutluOzel-b commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -351,51 +357,63 @@ protected void checkDuplicates(ParserRuleContext ctx, Map<Integer, ErrorSeverity
+   * @param rule Generic rule to check
+   * @param options Options checked to insert into error message
+   */
+  protected void checkHasIllegalOptions(TerminalNode rule, String options) {
```
Please remove this method

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,121 @@
+
+    private static final Map<Integer, ErrorSeverity> DUPLICATE_CHECK_OPTIONS = new HashMap<Integer, ErrorSeverity>() {
+        {
+            put(CICSLexer.STARTBROWSE, ErrorSeverity.ERROR);
```
`STARTBROWSE` is not an operand.

mm-broadcom replied:
Resolved in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    private void checkActivityIdProcess(CICSParser.Cics_startbrowse_activityid_processContext ctx) {
+        checkMutuallyExclusiveOptions("ACTIVITYID or PROCESS", ctx.ACTIVITYID(), ctx.PROCESS());
```
`PROCESSTYPE` requries `PROCESS`.

mm-broadcom replied:
Resolved in latest commit.

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -779,15 +779,16 @@ cics_zero_digit: LPARENCHAR ZERO_DIGIT RPARENCHAR;
+cics_startbrowse_container: (CONTAINER | cics_startbrowse_activityid_process | CHANNEL cics_data_value | (BROWSETOKEN | RETCODE) cics_data_area | cics_handle_response)+;
+cics_startbrowse_event: (EVENT | BROWSETOKEN cics_data_area | ACTIVITYID cics_data_value | cics_handle_response)+;
+cics_startbrowse_process: (PROCESS | PROCESSTYPE cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
+cics_startbrowse_timer: (TIMER cics_data_value | BROWSETOKEN cics_data_area | cics_handle_response)+;
```
I am unable to compile an example where the `TIMER` is actually followed by the `cics_data_value`. It looks like an error in the doc.

mm-broadcom replied:
Resolved in latest commit.

mm-broadcom committed:
Address issues brought up in PR

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -83,22 +82,22 @@ public <E extends ParserRuleContext> void checkOptions(E ctx) {
         }
     }
 
     private void checkActivity(CICSParser.Cics_startbrowse_activityContext ctx) {
```
`ACTIVITYID` and `PROCESS` are mutually exclusive as well.

mm-broadcom replied:
Fixed in latest commits.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -83,22 +82,22 @@ public <E extends ParserRuleContext> void checkOptions(E ctx) {
+        checkPrerequisiteIsMet(ctx.PROCESS(), ctx.PROCESSTYPE(), ctx, "PROCESSTYPE without PROCESS");
     }
 
     private void checkContainer(CICSParser.Cics_startbrowse_containerContext ctx) {
```
`ACTIVITYID`, `PROCESS` and `CHANNEL` are mutually exclusive as well.

mm-broadcom replied:
Fixed in latest commits.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,116 @@
+
+    private static final Map<Integer, ErrorSeverity> DUPLICATE_CHECK_OPTIONS = new HashMap<Integer, ErrorSeverity>() {
+        {
+            put(CICSLexer.ACTIVITYID, ErrorSeverity.ERROR);
```
`ACTIVITY`, `CONTAINER`, <del>`CHANNEL`</del>, `EVENT` and `TIMER` produce warnings.

mm-broadcom replied:
Fixed in latest commits.

slavek-kucera replied:
This got somehow reverted.
`ACTIVITY`, `CONTAINER`, `EVENT` and `TIMER` produce warnings.

mm-broadcom committed:
Fix duplication severity, added exclusivity checks

mm-broadcom committed:
More fixes

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,114 @@
+        checkHasMandatoryOptions(ctx.BROWSETOKEN(), ctx, "BROWSETOKEN");
+        checkPrerequisiteIsMet(ctx.PROCESS(), ctx.PROCESSTYPE(), ctx, "PROCESSTYPE without PROCESS");
+        checkMutuallyExclusiveOptions("ACTIVITYID, PROCESS or CHANNEL", ctx.ACTIVITYID(), ctx.PROCESS(), ctx.CHANNEL());
+    }
```
`CHANNEL` requires `CONTAINER`.

Also the more I look at this the more I am wondering if keeping the approach with 5 grammar rules for the options is a good idea given the similarity between them.

mm-broadcom committed:
Consolidate grammar, fix tests, change checks

mm-broadcom committed:
Fix tests to use variables.

slavek-kucera commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsExciStartBrowseStatement.java:
```diff
@@ -14,36 +14,107 @@
+ * https://www.ibm.com/docs/en/cics-ts/6.x?topic=summary-startbrowse-activity
  */
 public class TestCicsExciStartBrowseStatement {
-  public static final String TEXT =
```
This test should have been kept - the EXCI variant should still pass.

mm-broadcom replied:
Restored in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,102 @@
+            checkHasIllegalOptions(ctx.PROCESS(), "PROCESS");
+            checkHasIllegalOptions(ctx.PROCESSTYPE(), "PROCESSTYPE");
+        } else if (!ctx.PROCESS().isEmpty()) {
+            checkHasMandatoryOptions(ctx.PROCESSTYPE(), ctx, "PROCESSTYPE");
```
The PROCESS without an operand is allowed only in this case.

mm-broadcom committed:
Restore EXCI test

mm-broadcom committed:
Add PROCESS/PROCESS(value) differentiation

Add the ability to differentiate between PROCESS on its own and PROCESS with a value and warn accordingly. Add tests as well

mm-broadcom committed:
Fix check for PROCESS in mutualExclusive for CONTAINER subrule

mm-broadcom commented:
@slavek-kucera Can you review this?

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -120,27 +120,27 @@ protected boolean checkHasMandatoryOptions(List<?> rules, ParserRuleContext ctx,
      * @param options         - String of the element that is required.
      */
-    protected <E extends ParseTree> void checkPrerequisiteIsMet(List<E> requiredContext, List<E> optionalContext, ParserRuleContext ctx, String options) {
+    protected <E> void checkPrerequisiteIsMet(List<E> requiredContext, List<E> optionalContext, ParserRuleContext ctx, String options) {
```
Why was this needed? I think all the Es should actually be ParseTree(s)

mm-broadcom replied:
It was needed such that I can separate `PROCESS` (on its own) from `PROCESS cics_data_value` which used another rule to facilitate the difference. As the rule is not a ParseTree that change had to be made.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+            put(CICSLexer.ACTIVITY, ErrorSeverity.WARNING);
+            put(CICSLexer.BROWSETOKEN, ErrorSeverity.ERROR);
+            put(CICSLexer.CONTAINER, ErrorSeverity.WARNING);
+            put(CICSLexer.CHANNEL, ErrorSeverity.WARNING);
```
error.

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+            put(CICSLexer.CHANNEL, ErrorSeverity.WARNING);
+            put(CICSLexer.EVENT, ErrorSeverity.WARNING);
+            put(CICSLexer.TIMER, ErrorSeverity.WARNING);
+            put(CICSLexer.RETCODE, ErrorSeverity.ERROR);
```
RETCODE is not part of the grammar.

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+
+        if (ctx.PROCESS().isEmpty()) {
+            // PROCESS can be its own command outside the context of CONTAINER or as part of that one.
+            checkHasExactlyOneOption("ACTIVITY, CONTAINER, EVENT or TIMER", ctx, ctx.ACTIVITY(), ctx.CONTAINER(), ctx.EVENT(), ctx.TIMER());
```
Since the grammar now distinguishes between PROCESS and PROCESS(withargument), I think this check should include the PROCESS() list as well.

mm-broadcom replied:
That will not work. This check is if PROCESS is empty to avoid colliding with STARTBROWSE PROCESS: https://www.ibm.com/docs/en/cics-ts/6.x?topic=summary-startbrowse-process

slavek-kucera replied:
I don't follow. This should be just 
```
checkHasExactlyOneOption("ACTIVITY, CONTAINER, PROCESS, EVENT or TIMER", ctx, ctx.ACTIVITY(), ctx.CONTAINER(), ctx.PROCESS(), ctx.EVENT(), ctx.TIMER());
```
without the if statement.

mm-broadcom replied:
Added in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+
+    private void checkBody(CICSParser.Cics_startbrowse_bodyContext ctx) {
+        if (!ctx.ACTIVITY().isEmpty()) {
+            checkPrerequisiteIsMet(ctx.cics_startbrowse_processWithValue_subrule(), ctx.PROCESSTYPE(), ctx, "PROCESSTYPE without PROCESS");
```
all or nothing.

mm-broadcom replied:
Added functionality in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+        if (!ctx.ACTIVITY().isEmpty()) {
+            checkPrerequisiteIsMet(ctx.cics_startbrowse_processWithValue_subrule(), ctx.PROCESSTYPE(), ctx, "PROCESSTYPE without PROCESS");
+            checkHasIllegalOptions(ctx.PROCESS(), "PROCESS, in this context, requires a value");
+            checkMutuallyExclusiveOptions("ACTIVITYID or PROCESS", ctx.ACTIVITYID(), ctx.PROCESS());
```
cics_startbrowse_processWithValue_subrule instead of PROCESS

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+            checkPrerequisiteIsMet(ctx.cics_startbrowse_processWithValue_subrule(), ctx.PROCESSTYPE(), ctx, "PROCESSTYPE without PROCESS");
+            checkHasIllegalOptions(ctx.PROCESS(), "PROCESS, in this context, requires a value");
+            checkMutuallyExclusiveOptions("ACTIVITYID or PROCESS", ctx.ACTIVITYID(), ctx.PROCESS());
+        } else if (!ctx.CONTAINER().isEmpty()) {
```
if (channel) {
all other tokens like ACTIVITYID, PROCESS, etc. are illegal
}

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSStartbrowseOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+            checkHasIllegalOptions(ctx.PROCESSTYPE(), "PROCESSTYPE");
+        } else if (!ctx.PROCESS().isEmpty()) {
+            checkHasMandatoryOptions(ctx.PROCESSTYPE(), ctx, "PROCESSTYPE");
+            checkHasIllegalOptions(ctx.cics_startbrowse_processWithValue_subrule(), "PROCESS, in this context, cannot have a value");
```
ACTIVITYID
and
ACTIVITY 
CONTAINER 
EVENT 
CHANNEL
TIMER

are illegal here.

mm-broadcom committed:
Adjust duplicate severity, fix check, add check for CHANNEL

mm-broadcom committed:
Improve CHANNEL's check.

mm-broadcom committed:
Add illegal option checks to PROCESS

mm-broadcom committed:
Fix test to reflect new checks

mm-broadcom committed:
Improve checks and fix test

mm-broadcom committed:
Use AllOrNothing function for check, change duplicate severity

mm-broadcom committed:
more fixes

mm-broadcom committed:
Change PROCESS back to warning
